Sources/SwiftDiagnostics/Note.swift

```diff
@@ -0,0 +1,68 @@
+//===--- Note.swift -------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+
+/// Types conforming to this protocol represent note messages that can be
+/// shown to the client.
+/// The messages should describe what the note is pointing at.
+public protocol NoteMessage {
+  /// The message that should be displayed in the client.
+  var message: String { get }
+
+  /// See ``MessageID``.
+  var fixItID: MessageID { get }
+}
+
+/// A note that points to another node that's relevant for a Diagnostic.
+public struct Note: CustomDebugStringConvertible {
+  /// The node whose location the node is pointing.
+  public let node: Syntax
+
+  /// The position at which the location should be anchored.
+  /// By default, this is the start location of `node`.
+  public let position: AbsolutePosition
+
+  /// A description of what this note is pointing at.
+  public let noteMessage: NoteMessage
```

Comment:
I'm still not sure having an ID for note/diagnostic is really worth it. But we can keep it for now 🤷

Reply:
I still think it’s very little overhead at this point and once a use case comes up we will be glad that we have so we don’t need to do string matching for the error message.